lib/OpenQA/WebAPI/Controller/API/V1/JobTemplate.pm

```diff
@@ -236,151 +266,157 @@ sub update {
-                                last if (@$errors);
                             }
+                            $job_template_settings->search(
+                                {
+                                    id              => {'not in' => \@setting_ids},
+                                    job_template_id => $job_template_id,
+                                })->delete();
+
+                            # Stop iterating if there were errors with this test suite
+                            last if (@$errors);
                         }
                     }
+                }
 
-                    # Drop entries we haven't touched in add/update loop
-                    $job_templates->search(
-                        {
-                            id       => {'not in' => \@job_template_ids},
-                            group_id => $group_id,
-                        })->delete();
-
-                    # Preview mode: Get the expected YAML and rollback the result
-                    if ($self->param('preview')) {
-                        $json->{template} = YAML::XS::Dump($self->get_job_groups($json->{id}));
-                        $self->schema->txn_rollback;
-                    }
-                });
-        }
-        catch {
-            # Push the exception to the list of errors without the trailing new line
-            push @$errors, substr($_, 0, -1);
-        };
-
-        if (@$errors) {
-            $json->{error} = \@$errors;
-            $self->app->log->error(@$errors);
-            $self->respond_to(json => {json => $json, status => 400},);
-            return;
-        }
-
-        $self->emit_event('openqa_jobtemplate_create', $json) unless $self->param('preview');
-        $self->respond_to(json => {json => $json});
+                # Drop entries we haven't touched in add/update loop
+                $job_templates->search(
+                    {
+                        id       => {'not in' => \@job_template_ids},
+                        group_id => $group_id,
+                    })->delete();
+
+                # Preview mode: Get the expected YAML and rollback the result
+                if ($self->param('preview')) {
+                    $json->{template} = $self->get_job_groups($group_id)->{$job_group->name};
+                    $self->schema->txn_rollback;
+                }
+                else {
+                    # Store the original YAML template after all changes have been made
+                    $job_group->update({template => $self->param('template')});
+                }
+            });
+    }
+    catch {
+        # Push the exception to the list of errors without the trailing new line
+        push @$errors, substr($_, 0, -1);
+    };
 
-        # Process only one group
-        last;
+    if (@$errors) {
+        $json->{error} = \@$errors;
+        $self->app->log->error(@$errors);
+        $self->respond_to(json => {json => $json, status => 400},);
```

Comment:
That trailing comma doesn't belong there.

Reply:
It is annoying that tidy doesn't remove commas when turning multi-line lists into one liners.